app/controllers/single_page_subscriptions_controller.rb

```diff
@@ -0,0 +1,63 @@
+class SinglePageSubscriptionsController < ApplicationController
+  include AccountHelper
+  include GovukPersonalisation::ControllerConcern
+
+  before_action do
+    head :not_found unless single_page_notifications_feature_enabled?
+  end
+
+  def show
+    content_item = GdsApi.content_store.content_item(topic).to_h
+    if logged_in?
+      user_id = GdsApi.account_api.get_user(
+        govuk_account_session: @account_session_header,
+      ).to_h["id"]
+
+      begin
+        GdsApi.email_alert_api.find_subscriber_by_govuk_account(
+          govuk_account_id: user_id,
+        )
+      rescue GdsApi::HTTPNotFound
+        GdsApi.email_alert_api.link_subscriber_to_govuk_account(
+          govuk_account_session: @account_session_header,
+        )
+      ensure
+        subscriber_list_slug = GdsApi.email_alert_api.find_or_create_subscriber_list(
+          { url: topic, name: content_item["title"] },
+        ).to_h.dig("subscriber_list", "slug")
+
+        GdsApi.account_api.put_email_subscription(
+          name: content_item["title"],
+          topic_slug: subscriber_list_slug,
+          govuk_account_session: @account_session_header,
+        )
+
+        redirect_to topic
+      end
+
+    end
+  rescue GdsApi::ContentStore::ItemNotFound
+    head :not_found
+  rescue GdsApi::HTTPUnauthorized
+    logout!
+    redirect_with_analytics single_page_session_path
+  end
+
+  def edit
+    redirect_with_analytics single_page_session_path
+  end
+
+private
+
+  def topic
+    @topic ||= params[:topic]
+  end
+
+  def single_page_session_path
+    redirect_path = "/email/subscriptions/account/confirm?topic=#{params[:topic]}"
+
+    GdsApi.account_api.get_sign_in_url(
+      redirect_path: redirect_path,
+    )["auth_uri"] + "?redirect_path=#{redirect_path}"
```

Comment:
Why are you adding an extra `redirect_path` parameter here?

Reply:
We thought we'd need it so DI can return the user back here to try again after being logged in? Is that not the plan?

Reply:
To clarify, we weren't sure what `get_sign_in_url` returned, and [the stub](https://github.com/alphagov/gds-api-adapters/blob/master/lib/gds_api/test_helpers/account_api.rb#L22-L29) implies it comes back with an `auth_uri` that doesn't have the redirect param, so we added it on here.

Reply:
The stub comes back with whatever URI you want.  If you go to https://www.gov.uk you can see what URL you get redirected to to see what it's like:

https://oidc.account.gov.uk/authorize?client_id=LcueBVCnGZw-YFdTZ4S07XbQx7I&nonce=5cddc860bbe2d6ac6f753bfb8114de5d&redirect_uri=https%3A%2F%2Fwww.gov.uk%2Fsign-in%2Fcallback&response_type=code&scope=openid%20email%20govuk-account&state=1f2e57677c24b3458bc0dc1007dd50bf%3A312423&vtr=%5B%22Cl%22%2C%22Cl.Cm%22%5D

There's a `redirect_uri` parameter there which is what DI redirects us back to:

```
redirect_uri=https://www.gov.uk/sign-in/callback
```

Which is constructed from the `redirect_path` parameter in the `get_sign_in_url` call.